File: 7 - Recursion/Python/Recursion.py
# Define sum_to_one() below...
def sum_to_one(n):
  if n == 1:
    return n
  
  print("Recursing with input: {0}".format(n))
  return n + sum_to_one(n - 1)

def sum_digits(n):
  if n < 10:
    return n
  else:
    last_digit = n % 10
    # What argument is every digit except the last?
    return last_digit + sum_digits(n // 10)

  # Define factorial() below:
def factorial(n):
  if n < 2:
    return 1
  
  return n * factorial(n - 1)

def power_set_iter(set):
  power_set_size = 2**len(set)
  result = []
 
  for bit in range(0, power_set_size):
    sub_set = []
    for binary_digit in range(0, len(set)):
      if((bit & (1 << binary_digit)) > 0):
        sub_set.append(set[binary_digit])
    result.append(sub_set)
  return result

def power_set(my_list):
    # base case: an empty list
    if len(my_list) == 0:
        return [[]]
    # recursive step: subsets without first element
    power_set_without_first = power_set(my_list[1:])
    # subsets with first element
    with_first = [ [my_list[0]] + rest for rest in power_set_without_first ]
    # return combination of the two
    return with_first + power_set_without_first

# define flatten() below...
def flatten(my_list):
  result = []

  for element in my_list:
    if isinstance(element, list):
      print("List found!")

      flat_list = flatten(element)

      result = result + flat_list
    else:
      result.append(element)

  return result

# define the fibonacci() function below...
def fibonacci(n):
  #print("Called")
  if n == 1:
    return 1
  if n == 0:
    return 0
  return fibonacci(n - 2) + fibonacci(n - 1)

# Define build_bst() below...
def build_bst(my_list):
  if len(my_list) == 0:
    return "No Child"
  
  middle_idx = int(len(my_list) / 2)
  middle_value = my_list[middle_idx]

  print("Middle index: {0}".format(middle_idx))
  print("Middle value: {0}".format(middle_value))

  tree_node = {
    "data": middle_value
  }

  tree_node["left_child"] = build_bst(my_list[:middle_idx])
  tree_node["right_child"] = build_bst(my_list[middle_idx + 1:])

  return tree_node

### reserve for testing...
planets = ['mercury', 'venus', ['earth'], 'mars', [['jupiter', 'saturn']], 'uranus', ['neptune', 'pluto']]
universities = ['MIT', 'UCLA', 'Stanford', 'NYU', 'UofU']
power_set_of_universities = power_set(universities)

# For testing BST
sorted_list = [12, 13, 14, 15, 16]
binary_search_tree = build_bst(sorted_list)


print(sum_to_one(7))
sum_digits(12) # 3
sum_digits(194) # 14
print(factorial(4))
# Will stack overflow due to max recursion depth
# print(factorial(56151643515))

for set in power_set_of_universities:
  print(set)

print(flatten(planets))
print(fibonacci(4))
print(binary_search_tree)

File: 7 - Recursion/TypeScript/src/ts/Recursion.ts

//////////////////// Basic Examples ////////////////////

// O(n)
function sumToOne(n : number) : number
{
    if (n == 1)
    {
        return n;
    }

    return n + sumToOne(n - 1);
}

// O(n)
function sumDigits(n : number) : number
{
    if (n < 10)
    {
        return n;
    }

    return (n % 10) + sumDigits(Math.floor(n / 10));
}

// O(n)
function factorial(n : number) : number
{
    if (n == 0)
    {
        return 1;
    }

    return n * factorial(n - 1);
}

// O(n)
function flatten<T>(list : T[]) : T[]
{
    let result : T[] = [];

    for (let elem of list)
    {
        if (elem instanceof Array)
        {
            //console.log("List found");
            const flatList : T[] = flatten(elem);
            result = result.concat(flatList);
        }
        else
        {
            result.push(elem);
        }
    }

    return result;
}

// O(2^n)
// Can improve with memoization (Dynamic Programming), see last chapter (19 - Example Problems)
function fibonacci(n : number) : number
{
    if (n == 0)
    {
        return 0;
    }

    if (n == 1)
    {
        return 1;
    }

    return fibonacci(n - 2) + fibonacci(n - 1);
}

//////////////////// Advanced Examples ////////////////////

// Note: This shows how to do array types differently compared to the flatten() function, but they
// mean the same thing. I.e. T[] == Array<T> and T[][] == Array<Array<T>>
//
// O(2^n), we’ll never do better than that because a set of N elements creates a power set of 2^N elements.
function powerSet<T>(list : Array<T>) : Array<Array<T>>
{
    // base case: an empty list
    if (list.length == 0)
    {
        return [[]];
    }

    // recursive step

    // Splice out the first element so the list heads towards the base case in the recursive calls
    const firstElement = list.splice(0, 1);

    // subsets without first element
    const powerSetWithoutFirst = powerSet(list);
    // subsets with first element
    let withFirst : Array<Array<T>> = [];

    for (let elem of powerSetWithoutFirst)
    {
        withFirst.push(firstElement.concat(elem));
    }

    // return a combination of the two
    return withFirst.concat(powerSetWithoutFirst);
}

// Assumes the list passed in is sorted
// O(nlogn)
function buildBST(list : (string | number)[]) : (object | null)
{
    if (list.length == 0)
    {
        return null;
    }

    const middleIndex : number = Math.floor(list.length / 2);
    const middleValue : (string | number) = list[middleIndex];

    // For testing
    //console.log(`Middle Index: ${middleIndex}`);
    //console.log(`Middle Value: ${middleValue}`);

    let treeNode : object = {
        data: middleValue
    };

    let leftTree = list.filter((value : (string | number), index : number, arr : (string | number)[]) => {
        return index < middleIndex;
    });

    let rightTree = list.filter((value : (string | number), index : number, arr : (string | number)[]) => {
        return index > middleIndex;
    });;

    treeNode["leftChild"] = buildBST(leftTree);
    treeNode["rightChild"] = buildBST(rightTree);

    return treeNode;
}

// For testing the basic recursive functions

const planets = ['mercury', 'venus', ['earth'], 'mars', [['jupiter', 'saturn']], 'uranus', ['neptune', 'pluto']];

console.log();

console.log("sumToOne(4): " + sumToOne(4));

console.log();

console.log("sumDigits(12): " + sumDigits(12));
console.log("sumDigits(194): " + sumDigits(194));

console.log();

console.log("factorial(4): " + factorial(4));
// Will stack overflow due to the allowed max call stack size (recursion depth)
//console.log("factorial(56151643515): " + factorial(56151643515));

console.log();

console.log("Flatten:");

console.log();

console.log(flatten(planets));

console.log();

console.log("fibonnaci(5): " + fibonacci(5));
console.log("fibonnaci(6): " + fibonacci(6));

console.log();


// For testing powerSet

const universities : string[] = ['MIT', 'UCLA', 'Stanford', 'NYU', 'UofU'];
console.log("Power Set of: [" + universities + "]");
const powerSetOfUniversities : any[][] = powerSet(universities);

console.log();
console.log();

for (let set of powerSetOfUniversities)
{
    console.log(set);
}

// For testing buildBST

// buildBST expects a sorted list
const sortedList : number[] = [12, 13, 14, 15, 16];
const binarySearchTree : (object | null) = buildBST(sortedList);

console.log();

console.log("Build Binary Search Tree (BST) From: [" + sortedList + "]");

console.log();

console.log(binarySearchTree);